<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/451024427.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [2]:
DBNAME = ''
USER = ''
PASSWORD = ''
HOST = ''
PORT = 0

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) FROM vacancies
'''

In [5]:
# результат запроса
query_result = pd.read_sql_query(query_3_1, connection)
display(query_result)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/3054148448.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
employers_count = f'''
SELECT COUNT(*) FROM employers
'''

In [7]:
# результат запроса
query_result = pd.read_sql_query(employers_count, connection)
display(query_result)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2105510051.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(employers_count, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
areas_count = f'''
SELECT COUNT(*) FROM areas
'''

In [9]:
# результат запроса
areas_qr = pd.read_sql_query(areas_count, connection)
display(areas_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/100260981.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas_qr = pd.read_sql_query(areas_count, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
industries_count = f'''
SELECT COUNT(*) FROM industries
'''

In [11]:
# результат запроса
industries_qr = pd.read_sql_query(industries_count, connection)
display(industries_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2474625399.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industries_qr = pd.read_sql_query(industries_count, connection)


count
0    294

***

# Выводы по предварительному анализу данных
В базе данных содержится 49197 вакансий из 294 сфер деятельности, которые распределены между 23501 работодателями, находящимися в 1362 регионах.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
vacancies_count_by_area = f'''
SELECT a.name, COUNT(v.id) as vacancies_count
FROM vacancies v
JOIN areas a ON v.area_id = a.id
GROUP BY a.name
ORDER BY vacancies_count DESC
LIMIT 5
'''

In [14]:
# результат запроса
vacancies_count_by_area_qr = pd.read_sql_query(vacancies_count_by_area, connection)
display(vacancies_count_by_area_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2341054921.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_count_by_area_qr = pd.read_sql_query(vacancies_count_by_area, connection)


name  vacancies_count
0           Москва             5333
1  Санкт-Петербург             2851
2            Минск             2112
3      Новосибирск             2006
4           Алматы             1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
vacancies_with_salary = f'''
SELECT COUNT(*) FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [16]:
# результат запроса
vacancies_with_salary_qr = pd.read_sql_query(vacancies_with_salary, connection)
display(vacancies_with_salary_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/1083759566.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_with_salary_qr = pd.read_sql_query(vacancies_with_salary, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
avg_salaries = f'''
SELECT AVG(salary_from) as avg_salary_from, AVG(salary_to) as avg_salary_to
FROM vacancies
'''

In [18]:
# результат запроса
avg_salaries_qr = pd.read_sql_query(avg_salaries, connection)
display(avg_salaries_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/1052188869.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salaries_qr = pd.read_sql_query(avg_salaries, connection)


avg_salary_from  avg_salary_to
0     71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
amount_of_vacancies_by_employment_and_schedule = f'''
SELECT v.employment, v.schedule, COUNT(v.id) as vacancies_count
FROM vacancies v
GROUP BY v.employment, v.schedule
ORDER BY vacancies_count DESC
'''

In [20]:
# результат запроса
amount_of_vacancies_by_employment_and_schedule_qr = pd.read_sql_query(amount_of_vacancies_by_employment_and_schedule, connection)
display(amount_of_vacancies_by_employment_and_schedule_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2236772814.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  amount_of_vacancies_by_employment_and_schedule_qr = pd.read_sql_query(amount_of_vacancies_by_employment_and_schedule, connection)


employment          schedule  vacancies_count
0      Полная занятость       Полный день            35367
1      Полная занятость  Удаленная работа             7802
2      Полная занятость     Гибкий график             1593
3   Частичная занятость  Удаленная работа             1312
4      Полная занятость    Сменный график              940
5            Стажировка       Полный день              569
6      Полная занятость    Вахтовый метод              367
7   Частичная занятость       Полный день              347
8   Частичная занятость     Гибкий график              312
9      Проектная работа       Полный день              141
10     Проектная работа  Удаленная работа              133
11           Стажировка     Гибкий график              116
12  Частичная занятость    Сменный график              101
13           Стажировка  Удаленная работа               64
14     Проектная работа     Гибкий график               18
15           Стажировка    Сменный график               12
16     Проектная работа    Вахтовый метод                2
17     Проектная работа    Сменный график                1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
amount_of_vacancies_by_experience = f'''
SELECT v.experience, COUNT(v.id) as vacancies_count
FROM vacancies v
GROUP BY v.experience
ORDER BY vacancies_count DESC
'''

In [22]:
# результат запроса
amount_of_vacancies_by_experience_qr = pd.read_sql_query(amount_of_vacancies_by_experience, connection)
display(amount_of_vacancies_by_experience_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/407631779.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  amount_of_vacancies_by_experience_qr = pd.read_sql_query(amount_of_vacancies_by_experience, connection)


experience  vacancies_count
0  От 1 года до 3 лет            26152
1       От 3 до 6 лет            14511
2           Нет опыта             7197
3         Более 6 лет             1337

***

# Выводы по детальному анализу вакансий
Мы можем видеть, что приблизительно у половины вакансий заполнено хотя бы одно поле с зарплатой, в остальных случаях эти поля не указаны. 

Самое большое количество вакансий находится в столицах (Москва, Минск, Алматы), при этом среди них также выделяются города миллионники (Санкт-Петербург, Новосибирск)

Средние значения заработной платы находятся в диапазоне 71064.66 - 110536.74

Большее количество работников требуется на Полную занятость и Полный рабочий день: 35367. При этом выделяется количество вакансий для опыта от 1 до 3 лет: 26152

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
# текст запроса
amount_of_vacancies_by_employer = f'''
SELECT e.name, COUNT(v.id) as vacancies_count
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY vacancies_count DESC
'''

In [25]:
# результат запроса
amount_of_vacancies_by_employer_qr = pd.read_sql_query(amount_of_vacancies_by_employer, connection)
display(amount_of_vacancies_by_employer_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/3031542727.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  amount_of_vacancies_by_employer_qr = pd.read_sql_query(amount_of_vacancies_by_employer, connection)


name  vacancies_count
0                      Яндекс             1933
1                  Ростелеком              491
2                    Тинькофф              444
3                        СБЕР              428
4               Газпром нефть              331
...                       ...              ...
14761            им. Тьюринга                1
14762        LLC Novalab Tech                1
14763   Компания Мир упаковки                1
14764  Лифинцев Олег Иванович                1
14765      ТрансЭнергоХолдинг                1

[14766 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
# текст запроса
amount_of_vacancies_in_areas = f'''
SELECT
    a.name,
    COUNT(e.id) AS number_of_employers,
    COUNT(v.id) AS number_of_vacancies
FROM areas a
LEFT JOIN employers e ON a.id = e.area
LEFT JOIN vacancies v ON a.id = v.area_id
GROUP BY a.id, a.name
HAVING COUNT(v.id) <= 1
ORDER BY number_of_employers DESC
LIMIT 1
'''

In [27]:
# результат запроса
amount_of_vacancies_in_areas_qr = pd.read_sql_query(amount_of_vacancies_in_areas, connection)
display(amount_of_vacancies_in_areas_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2879576979.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  amount_of_vacancies_in_areas_qr = pd.read_sql_query(amount_of_vacancies_in_areas, connection)


name  number_of_employers  number_of_vacancies
0  Россия                  410                    0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
# текст запроса
count_regions_where_employers_have_vacancies = f'''
SELECT e.id, e.name, COUNT(DISTINCT a.id) as regions_count 
FROM employers e
LEFT JOIN vacancies v ON e.id = v.employer_id
LEFT JOIN areas a ON v.area_id = a.id
GROUP BY e.id, e.name
ORDER BY regions_count DESC
'''

In [29]:
# результат запроса
count_regions_where_employers_have_vacancies_qr = pd.read_sql_query(count_regions_where_employers_have_vacancies, connection)
display(count_regions_where_employers_have_vacancies_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2768000854.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_regions_where_employers_have_vacancies_qr = pd.read_sql_query(count_regions_where_employers_have_vacancies, connection)


id                    name  regions_count
0         1740                  Яндекс            181
1         2748              Ростелеком            152
2      5724811              Спецремонт            116
3      5130287  Поляков Денис Иванович             88
4      3682876               ООО ЕФИН              71
...        ...                     ...            ...
23496  5287515              Но Смокинг              0
23497  2839354        МАСТЕРСКАЯ CREED              0
23498  5774578           Dzing Finance              0
23499  5775069                    КОРС              0
23500       15             Наука-Связь              0

[23501 rows x 3 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
# текст запроса
count_of_employers_without_industries = f'''
SELECT COUNT(e.id) as employers_count
FROM employers e
LEFT JOIN employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL
'''

In [31]:
# результат запроса
count_of_employers_without_industries_qr = pd.read_sql_query(count_of_employers_without_industries, connection)
display(count_of_employers_without_industries_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2170412005.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_employers_without_industries_qr = pd.read_sql_query(count_of_employers_without_industries, connection)


employers_count
0             8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
# текст запроса
company_name_with_four_industries = f'''
SELECT e.name, COUNT(DISTINCT ei.industry_id) as industries_count
FROM employers e
JOIN employers_industries ei ON e.id = ei.employer_id
GROUP BY e.name
HAVING COUNT(DISTINCT ei.industry_id) = 4
ORDER BY e.name
'''

In [33]:
# результат запроса
company_name_with_four_industries_qr = pd.read_sql_query(company_name_with_four_industries, connection)
display(company_name_with_four_industries_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2556726634.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  company_name_with_four_industries_qr = pd.read_sql_query(company_name_with_four_industries, connection)


name  industries_count
0        101 Интернет                 4
1            21vek.by                 4
2                2ГИС                 4
3                  2К                 4
4         4 пикселя +                 4
...               ...               ...
1127            ЮРИОН                 4
1128  ЮТИП Технологии                 4
1129           ЯКласс                 4
1130            ЯрНео                 4
1131    Яшин&Партнёры                 4

[1132 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
count_employers_with_industry_development = f'''
SELECT COUNT(DISTINCT e.id) as employers_count, i.name as industry_name
FROM employers e
JOIN employers_industries ei ON e.id = ei.employer_id
JOIN industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения'
GROUP BY i.name
'''

In [35]:
# результат запроса
count_employers_with_industry_development_qr = pd.read_sql_query(count_employers_with_industry_development, connection)
display(count_employers_with_industry_development_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2208603714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_employers_with_industry_development_qr = pd.read_sql_query(count_employers_with_industry_development, connection)


employers_count                        industry_name
0             3553  Разработка программного обеспечения

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников
million_cities_page = pd.read_html('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')
table = million_cities_page[0]
display(table)

cities_list = table['Город'].tolist()
display(cities_list)

№            Город Население, тыс. чел.  (на 1 января 2024)[2]  \
0    1           Москва                                      13 150   
1    2  Санкт-Петербург                                       5 598   
2    3      Новосибирск                                       1 634   
3    4     Екатеринбург                                       1 536   
4    5           Казань                                       1 319   
5    6       Красноярск                                       1 205   
6    7  Нижний Новгород                                       1 205   
7    8        Челябинск                                       1 177   
8    9              Уфа                                       1 163   
9   10           Самара                                       1 159   
10  11   Ростов-на-Дону                                       1 140   
11  12        Краснодар                                       1 139   
12  13             Омск                                       1 104   
13  14          Воронеж                                       1 046   
14  15            Пермь                                       1 027   
15  16        Волгоград                                       1 019   

   Население, тыс. чел.  (перепись-2021)[3]  \
0                                    13 010   
1                                     5 602   
2                                     1 634   
3                                     1 544   
4                                     1 309   
5                                     1 188   
6                                     1 226   
7                                     1 190   
8                                     1 145   
9                                     1 173   
10                                    1 142   
11                                    1 099   
12                                    1 126   
13                                    1 058   
14                                    1 034   
15                                    1 028   

   Прирост  относительно  переписи-2010  \
0                                13,1 %   
1                                14,8 %   
2                                10,9 %   
3                                14,4 %   
4                                14,4 %   
5                                21,6 %   
6                                −2,0 %   
7                                 5,3 %   
8                                 7,8 %   
9                                 0,3 %   
10                                4,9 %   
11                               43,8 %   
12                               −3,4 %   
13                               18,9 %   
14                                4,3 %   
15                                0,7 %   

   Население, тыс. чел. (перепись-2010)[4]  
0                                   11 504  
1                                    4 880  
2                                    1 474  
3                                    1 350  
4                                    1 144  
5                                      977  
6                                    1 251  
7                                    1 130  
8                                    1 062  
9                                    1 169  
10                                   1 089  
11                                     764  
12                                   1 166  
13                                     890  
14                                     991  
15                                   1 021

['Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Самара',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград']

In [37]:
# текст запроса
count_yandex_vacancies_in_million_cities = f'''
SELECT a.name as city_name, COUNT(v.id) as vacancies_count
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
JOIN areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс'
AND a.name IN {tuple(cities_list)}
GROUP BY a.name
UNION
SELECT 'Total', COUNT(v.id) as vacancies_count
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
JOIN areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс'
AND a.name IN {tuple(cities_list)}
'''

In [38]:
# результат запроса
count_yandex_vacancies_in_million_cities_qr = pd.read_sql_query(count_yandex_vacancies_in_million_cities, connection)
display(count_yandex_vacancies_in_million_cities_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/662831893.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_yandex_vacancies_in_million_cities_qr = pd.read_sql_query(count_yandex_vacancies_in_million_cities, connection)


city_name  vacancies_count
0             Total              485
1         Волгоград               24
2           Воронеж               32
3      Екатеринбург               39
4            Казань               25
5         Краснодар               30
6        Красноярск               23
7            Москва               54
8   Нижний Новгород               36
9       Новосибирск               35
10             Омск               21
11            Пермь               25
12   Ростов-на-Дону               25
13           Самара               26
14  Санкт-Петербург               42
15              Уфа               26
16        Челябинск               22

***

# Выводы по анализу работодателей
Компания Яндекс является лидером по количеству вакансий - 1933 и по количеству регионов в которых публикуются вакансии компании - 181.

У 8419 не указана сфера деятельности, что может указывать на то, что компании занимаются несколькими сферами, либо существует проблема на этапе заполнения данных компании

В списке компаний находится 3553 компании со сферой деятельности "Разработка программного обеспечения"

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [58]:
# текст запроса
vacancies_count_data_related = f'''
SELECT COUNT(v.id) as vacancies_count
FROM vacancies v
WHERE LOWER(v.name) LIKE '%data%' OR LOWER(v.name) LIKE '%данн%'
'''

In [59]:
# результат запроса
vacancies_count_data_related_qr = pd.read_sql_query(vacancies_count_data_related, connection)
display(vacancies_count_data_related_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/1969848728.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_count_data_related_qr = pd.read_sql_query(vacancies_count_data_related, connection)


vacancies_count
0             1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [144]:
# текст запроса
data_science_vacancy_base_query = f'''
FROM vacancies v
WHERE (LOWER(v.name) LIKE '%data scientist%'
      OR LOWER(v.name) LIKE '%data science%'
      OR LOWER(v.name) LIKE '%исследователь данных%'
      OR v.name LIKE '%ML%'
      OR LOWER(v.name) LIKE '%machine learning%'
      OR LOWER(v.name) LIKE '%машинн%обучен%')
      AND v.name NOT ILIKE '%html%'
'''

junior_data_science_vacancies = f'''
SELECT COUNT(v.id) as vacancies_count
{data_science_vacancy_base_query}
      AND (LOWER(v.name) LIKE '%junior%'
        OR LOWER(v.experience) = 'нет опыта'
        OR LOWER(v.employment) = 'стажировка')
'''

In [100]:
# результат запроса
junior_data_science_vacancies_qr = pd.read_sql_query(junior_data_science_vacancies, connection)
display(junior_data_science_vacancies_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/1298482060.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  junior_data_science_vacancies_qr = pd.read_sql_query(junior_data_science_vacancies, connection)


vacancies_count
0               51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [101]:
# текст запроса
data_science_vacancies_with_SQL_or_Postgres = f'''
SELECT COUNT(v.id) as vacancies_count
{data_science_vacancy_base_query}
      AND (LOWER(v.key_skills) LIKE '%sql%'
        OR LOWER(v.key_skills) LIKE '%postgres%')
'''

In [102]:
# результат запроса
data_science_vacancies_with_SQL_or_Postgres_qr = pd.read_sql_query(data_science_vacancies_with_SQL_or_Postgres, connection)
display(data_science_vacancies_with_SQL_or_Postgres_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/3775525707.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_science_vacancies_with_SQL_or_Postgres_qr = pd.read_sql_query(data_science_vacancies_with_SQL_or_Postgres, connection)


vacancies_count
0              229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [94]:
# текст запроса
ds_with_python = f'''
SELECT COUNT(v.id) as vacancies_count
{data_science_vacancy_base_query}
      AND LOWER(v.key_skills) LIKE '%python%'
'''

In [95]:
# результат запроса
ds_with_python_count_qr = pd.read_sql_query(ds_with_python, connection)
display(ds_with_python_count_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/3859551901.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_with_python_count_qr = pd.read_sql_query(ds_with_python, connection)


vacancies_count
0              357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [145]:
# текст запроса
ds_key_skills_count = f'''
SELECT AVG(skills_count) as avg_skills_count
FROM (
    SELECT (LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1) as skills_count
    {data_science_vacancy_base_query}
    AND key_skills IS NOT NULL AND key_skills != ''
) as subquery;
'''

In [146]:
# результат запроса
ds_key_skills_count_qr = pd.read_sql_query(ds_key_skills_count, connection)
display(ds_key_skills_count_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/1313887414.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_key_skills_count_qr = pd.read_sql_query(ds_key_skills_count_two, connection)


avg_skills_count
0          6.406032

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [151]:
# текст запроса
slaries_for_ds_by_experience = f'''
SELECT v.experience, 
       ROUND(AVG(
            CASE
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
                WHEN salary_from IS NOT NULL THEN salary_from
                ELSE salary_to
            END
        )) as avg_salary
{data_science_vacancy_base_query}
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY v.experience
'''


In [152]:
# результат запроса
salaries_for_ds_by_experience_qr = pd.read_sql_query(slaries_for_ds_by_experience, connection)
display(salaries_for_ds_by_experience_qr)

/var/folders/gr/q4qv3vrj7yv8w1_w60f1j11m0000gn/T/ipykernel_25573/2623212353.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salaries_for_ds_by_experience_qr = pd.read_sql_query(slaries_for_ds_by_experience, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

# Выводы по предметному анализу
Вакансии по Data Science имеют высокую среднюю заработную плату, даже для начального уровня. В среднем, на позиции Data Scientist требуется владеть 6 навыками. Количество вакансий, имеющих отношение к данным - 1771, при этом количество вакансий в которых указан Python - 357, что скорее всего говорит о указании Python в описании вакансии, так как Python намного популярнее R.

# Общий вывод по проекту

# Итог исследования, обобщите выводы
1.	Зарплаты: Примерно у половины вакансий указана хотя бы одна из зарплатных составляющих, в остальных случаях эти данные отсутствуют.
2.	Географическое распределение: Наибольшее количество вакансий сосредоточено в столицах (Москва, Минск, Алматы), а также в крупных городах-миллионниках (Санкт-Петербург, Новосибирск).
3.	Зарплатный диапазон: Средние значения заработной платы колеблются в пределах от 71064.66 до 110536.74.
4.	Занятость и опыт: Большинство вакансий предполагает полную занятость и полный рабочий день (35367 вакансий). Наибольшее количество вакансий требует опыта работы от 1 до 3 лет (26152 вакансий).
5.	Лидеры рынка: Компания Яндекс лидирует по числу вакансий (1933) и по количеству регионов, в которых публикуются её вакансии (181).
6.	Проблемы в заполнении данных: У 8419 вакансий не указана сфера деятельности, что может указывать на многопрофильность компаний или проблемы при заполнении данных.
7.	Разработка ПО: В списке работодателей 3553 компании указали свою сферу деятельности как “Разработка программного обеспечения”.
8.	Data Science: Вакансии в области Data Science предлагают высокие средние зарплаты даже для начального уровня. В среднем, на позиции Data Scientist требуется владеть 6 навыками. Всего 1771 вакансий связаны с данными, из которых в 357 указано знание Python



In [ ]:
connection.close()